packages/medusa/src/services/discount-condition.ts

```diff
@@ -0,0 +1,233 @@
+          `DiscountCondition with id: ${conditionId} was not found`
+        )
+      }
+
+      return condition
+    })
+  }
+
+  protected resolveConditionType_(data: UpsertDiscountConditionInput):
+    | {
+        type: DiscountConditionType
+        resource_ids: string[]
+      }
+    | undefined {
+    switch (true) {
+      case !!data.products?.length:
+        return {
+          type: DiscountConditionType.PRODUCTS,
+          resource_ids: data.products!,
+        }
+      case !!data.product_collections?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_COLLECTIONS,
+          resource_ids: data.product_collections!,
+        }
+      case !!data.product_types?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TYPES,
+          resource_ids: data.product_types!,
+        }
+      case !!data.product_tags?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TAGS,
+          resource_ids: data.product_tags!,
+        }
+      case !!data.customer_groups?.length:
+        return {
+          type: DiscountConditionType.CUSTOMER_GROUPS,
+          resource_ids: data.customer_groups!,
+        }
+      default:
+        return undefined
+    }
+  }
+
+  async upsertCondition(
+    discountIdOrDiscount: string | Discount,
+    data: UpsertDiscountConditionInput
+  ): Promise<void> {
+    let resolvedConditionType
+
+    return await this.atomicPhase_(
+      async (manager: EntityManager) => {
+        resolvedConditionType = this.resolveConditionType_(data)
+
+        if (!resolvedConditionType) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Missing one of products, collections, tags, types or customer groups in data`
+          )
+        }
+
+        const discountConditionRepo: DiscountConditionRepository =
+          manager.getCustomRepository(this.discountConditionRepository_)
+        const discountRepo: DiscountRepository = manager.getCustomRepository(
+          this.discountRepository_
+        )
+
+        let discount
+        if (typeof discountIdOrDiscount === `string`) {
+          discount = await discountRepo.findOne({
+            where: { id: discountIdOrDiscount },
+            relations: ["rule"],
+          })
+
+          if (!discount) {
+            throw new MedusaError(
+              MedusaError.Types.NOT_FOUND,
+              `Discount with ${discountIdOrDiscount} was not found`
+            )
+          }
+        } else {
+          discount = discountIdOrDiscount
+
+          // ensure discount has rule_id, will also ensure that object is in fact discount
+          if (!discount?.rule_id) {
+            throw new MedusaError(
+              MedusaError.Types.INVALID_DATA,
+              `Discount is missing rule_id`
+            )
+          }
+        }
+
+        if (data.id) {
+          // throw if condition does not exist
+          const condition = await this.retrieve(data.id)
```

Comment:
as mention in a previous comment, one way to ensure that the condition belongs to the discount would be to retrieve it by inner joining the discount. That way, we are looking for a condition in the context of a discount